### PR TITLE
Adds Item Dispensers

### PR DESCRIPTION
**Pull request by Wallemations**
I remembered how horrible my first attempt at adding to this codebase went so I wanted to try redeeming myself by porting something I made myself, entirely modular.
## Okay but what is it
Item Dispensers, they're like stationary boxes built specifically for walls. Only accepts a single type of item _(and any of its children, but that's fun so I kept it)_.
They're great for reducing that nasty box clutter from your precious tables, racks, and shelves.
### Here is an example of what you can expect with item dispensers:

https://user-images.githubusercontent.com/66052067/141048985-24ee6b4e-dbba-4837-9b23-d133d5f8e614.mp4

The pre-made ones come in 7 flavors...

And there's even a wallmount for an empty one for you to fill up yourself!

https://user-images.githubusercontent.com/66052067/141049124-e9410b29-70d2-4ca7-833c-43b7264f464b.mp4

Clearly a server without item dispensers is a server that's missing out.
So get item dispensers today!

### CODE REVIEW

**Comment by CRITAWAKETS** [comment]
https://youtu.be/IJDgkrtUlTc

File: jollystation_modules/code/game/objects/structures/item_dispensers.dm [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+		STR.set_holdable(list(stock))
+		STR.max_combined_w_class = 30
+
+/obj/structure/item_dispenser/attackby(obj/item/I, mob/user, params)
+	if(istype(I, stock))
+		if(contents.len < charges)
+			playsound(loc, 'sound/machines/click.ogg', 15, TRUE, -3)
+			SEND_SIGNAL(src, COMSIG_TRY_STORAGE_INSERT, I, user)
+			contents += I
+			balloon_alert(user, "You insert [I] into [src].")
```
Comment (MrMelbert) [inline jollystation_modules/code/game/objects/structures/item_dispensers.dm:82]: Balloon alerts are typically un-capitalized, concise, and un-punctuated and 
```suggestion
			balloon_alert(user, "inserted [I]")
```

File: jollystation_modules/code/game/objects/structures/item_dispensers.dm [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+			if(contents.len == 1)
+				update_icon(UPDATE_OVERLAYS)
+			return
+		else
+			to_chat(user, span_notice("You can't fit more [item_name]\s in [src]!"))
```
Comment (MrMelbert) [inline jollystation_modules/code/game/objects/structures/item_dispensers.dm:87]:
```suggestion
			to_chat(user, span_warning("You can't fit [item_name] into [src]!"))
```

File: jollystation_modules/code/game/objects/structures/item_dispensers.dm [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		return
+	if(!stocked)
+		to_chat(user, span_notice("[src] hasn't been stocked yet!"))
+		return
+	var/obj/item/grabbies = locate(stock) in contents
+	if(grabbies && contents.len > 0)
+		SEND_SIGNAL(src, COMSIG_TRY_STORAGE_TAKE, grabbies, user)
+		user.put_in_hands(grabbies)
+		contents -= grabbies
+		balloon_alert(user, "You take \a [item_name] from [src]")
```
Comment (MrMelbert) [inline jollystation_modules/code/game/objects/structures/item_dispensers.dm:151]:
```suggestion
		balloon_alert(user, "removed [item_name]")
```

**Review by MrMelbert — COMMENTED** [review]
Looks pretty cool

File: jollystation_modules/code/game/objects/structures/item_dispensers.dm (hunk 1 above)
Comment (Wallemations) [reply jollystation_modules/code/game/objects/structures/item_dispensers.dm:82]: Ah, good to know, I'll be sure to fix that in the morning

**Review by Jolly-66 — COMMENTED** [review]
If the sprites are from goon put them in the goon folder please

**Comment by Jolly-66** [comment]
When I get home today I'm dropping a few more comments

Fuck mobile reviews

**Comment by Wallemations** [comment]
> If the sprites are from goon put them in the goon folder please 

Do I put the custom ones in the goon folder too or just the ones directly from goon

**Comment by Jolly-66** [comment]
> Do I put the custom ones in the goon folder too or just the ones directly from goon

If you made sprites keep them separate but since Goon has a different copyright than TG, I don't wanna get in trouble with them, even tho their chances of looking here are slim to none, I want to respect what TG does.

Also, per the rules, if you can, please make sure to credit the OG spriters.

File: jollystation_modules/code/game/objects/structures/item_dispensers.dm [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		playsound(loc, 'sound/machines/click.ogg', 15, TRUE, -3)
+		if(contents.len <= 0)
+			update_icon(UPDATE_OVERLAYS)
+	else
+		to_chat(user, span_notice("There are no [item_name]\s left in [src]."))
+
+//////      Dispensers      //////
+//////////////////////////////////
+//////  Pre-set Dispensers  //////
```
Comment (Jolly-66) [inline jollystation_modules/code/game/objects/structures/item_dispensers.dm:160]:
```suggestion
//Pre-set Dispensers
```

File: jollystation_modules/code/game/objects/structures/item_dispensers.dm [hunk 5]
```diff
@@ -0,0 +1,209 @@
+/*
+		Original sprites from Goonstation, the code was all me.
+		I saw the sprites and felt an irresistible urge to code them in myself.
+*/
+
+//////////////////////////////////
+//////      Dispensers      //////
```
Comment (Jolly-66) [inline jollystation_modules/code/game/objects/structures/item_dispensers.dm:7]:
```suggestion
//Dispensers, sprites from Goon.
```

File: jollystation_modules/code/game/objects/structures/item_dispensers.dm [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	stocked = TRUE
+
+/obj/structure/item_dispenser/bodybag
+	icon_state = "dispenser_bodybag"
+	stock = /obj/item/bodybag
+	stocked = TRUE
+
+////////  Pre-set Dispensers  /////////
+///////////////////////////////////////
+/// Wallframe for custom dispensers ///
```
Comment (Jolly-66) [inline jollystation_modules/code/game/objects/structures/item_dispensers.dm:200]:
```suggestion
//Preset Dispensers, Wallframe version
```

**Review by Jolly-66 — CHANGES_REQUESTED** [review]
I personally hate brick comments, just simplify them down and we'll be fine.

Clean code is happy code!

**Comment by Wallemations** [comment]
Git didn't remove the merge conflict tag automatically even though it's fixed
This is so sad

**Comment by Jolly-66** [comment]
> Git didn't remove the merge conflict tag automatically even though it's fixed This is so sad

Don't worry about it

**Review by Jolly-66 — APPROVED** [review]
:+1:

Unless @MrMelbert has any final comments

**Comment by Wallemations** [comment]
Hold on just found out `plural_s()` is a proc

**Pull request merged**